### PR TITLE
Improve handling of constant join conditions

### PR DESCRIPTION
**Pull request by mkleen**
## Summary of the changes / Why this improves CrateDB

See individual commits.

Fixes https://github.com/crate/crate/issues/11404 https://github.com/crate/crate/issues/12357

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by mkleen** [comment]
retest this please

unrelated test failure
```
REPRODUCE WITH: ./gradlew :server:test -Dtests.seed=AE399D67FD6FEDA7 -Dtests.class=io.crate.integrationtests.MetadataTrackerITest -Dtests.method="test_deleted_and_recreated_partition_is_also_deleted_and_restored_on_subscriber" -Dtests.locale=pa-PK -Dtests.timezone=Asia/Jakarta
```

File: server/src/test/java/io/crate/planner/operators/JoinTest.java [hunk 1]
```diff
@@ -245,7 +245,8 @@ public void testHashJoin_TableOrderInLogicalAndExecutionPlan() {
 
         LogicalPlan operator = createLogicalPlan(mss, tableStats);
         assertThat(operator, instanceOf(HashJoin.class));
-        assertThat(((HashJoin) operator).concreteRelation.toString(), is("DocTableRelation{doc.locations}"));
+        // smaller table must be on the right-hand-side
+        assertThat(((HashJoin) operator).rhs().getRelationNames(), contains(TEST_DOC_LOCATIONS_TABLE_IDENT));
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/planner/operators/JoinTest.java:249]:
```suggestion
        assertThat("smaller table must be on the right-hand-side", ((HashJoin) operator).rhs().getRelationNames(), contains(TEST_DOC_LOCATIONS_TABLE_IDENT));
```

File: server/src/test/java/io/crate/planner/operators/JoinTest.java [hunk 2]
```diff
@@ -265,7 +266,8 @@ public void testHashJoin_TablesSwitchWhenRightBiggerThanLeft() {
 
         LogicalPlan operator = createLogicalPlan(mss, tableStats);
         assertThat(operator, instanceOf(HashJoin.class));
-        assertThat(((HashJoin) operator).concreteRelation.toString(), is("DocTableRelation{doc.locations}"));
+        // smaller table must be on the right-hand-side
+        assertThat(((HashJoin) operator).rhs.getRelationNames(), contains(TEST_DOC_LOCATIONS_TABLE_IDENT));
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/planner/operators/JoinTest.java:270]:
```suggestion
        // smaller table must be on the right-hand-side
        assertThat("smaller table must be on the right-hand-side", ((HashJoin) operator).rhs.getRelationNames(), contains(TEST_DOC_LOCATIONS_TABLE_IDENT));
```

**Review by mfussenegger — APPROVED** [review]

**Comment by BaurzhanSakhariev** [comment]
is https://github.com/crate/crate/issues/12357 to be resolved by this PR?

**Comment by mkleen** [comment]
> is #12357 to be resolved by this PR?


Oh yes, it is. It's referenced in the code but not in the pr. Thank you for finding this!

**Comment by mkleen** [comment]
retest this please

unrelated test failure 

```
FAIL: test_graceful_stop_concurrent_queries (test_decommission.TestGracefulStopDuringQueryExecution)
```

**Pull request merged**